aioxmpp/node.py

```diff
@@ -1132,6 +1167,154 @@ def connected(self, *, presence=structs.PresenceState(False), **kwargs):
+        :return: IQ response :attr:`~.IQ.payload` or :data:`None`
+
+        Send the stanza and wait for it to be sent. If the stanza is an IQ
+        request, the response is awaited and the :attr:`~.IQ.payload` of the
+        response is returned.
+
+        If the stream is currently not ready, this method blocks until the
+        stream is ready to send payload stanzas. Note that this may be before
+        initial presence has been sent. To synchronise with that type of events,
+        use the appropriate signals.
+
+        The `timeout` as well as any of the exception cases referring to a
+        "response" do not apply for IQ response stanzas, message stanzas or
+        presence stanzas sent with this method, as this method only waits for
+        a reply if an IQ *request* stanza is being sent.
+
+        If `stanza` is an IQ request and the response is not received within
+        `timeout` seconds, :class:`TimeoutError` (not
+        :class:`asyncio.TimeoutError`!) is raised.
+
+        If `cb` is given, `stanza` must be an IQ request (otherwise,
+        :class:`ValueError` is raised before the stanza is sent). It must be a
+        callable returning an awaitable. It receives the response stanza as
+        first and only argument. The returned awaitable is awaited by
+        :meth:`send` and the result is returned instead of the original
+        payload. `cb` is called synchronously from the stream handling loop when
+        the response is received, so it can benefit from the strong ordering
+        guarantees given by XMPP XML Streams.
+
+        The `cb` may also return :data:`None`, in which case :meth:`send` will
+        simply return the IQ payload as if `cb` was not given. Since the return
+        value of coroutine functions is awaitable, it is valid and supported to
+        pass a coroutine function as `cb`.
+
+        .. warning::
+
+            Remember that it is an implementation detail of the event loop when
+            a coroutine is scheduled after it awaited an awaitable; this implies
+            that if the caller of :meth:`send` is merely awaiting the
+            :meth:`send` coroutine, the strong ordering guarantees of XMPP XML
+            Streams are lost.
+
+            To regain those, use the `cb` argument.
+
+        .. note::
+
+            For the sake of readability, unless you really need the strong
+            ordering guarantees, avoid the use of the `cb` argument. Avoid
+            using a coroutine function unless you really need to.
+
+        .. versionchanged:: 0.10
+
+            * This method now waits until the stream is ready to send stanza¸
+              payloads.
+            * This method was moved from
+              :meth:`aioxmpp.stream.StanzaStream.send`.
+
+        .. versionchanged:: 0.9
+
+            The `cb` argument was added.
+
+        .. versionadded:: 0.8
+        """
+        stopped_fut = self.on_stopped.future()
```

Comment:
This requires an extra scheduling cycle for every send! It may be better to check the state of the established_event and only construct the futures and wait for them it if established_event is not set!

Reply:
Good point!